File: Chapter 1/5. Graphical Editor/pc165.py
from sys import stdin

#from pandas import DataFrame

# x left ro right, y top to bottom
#test to see if max min needed
class Image:
    
    def __init__(self):
        self.img = []
        
    def command(self, cmd):
        if cmd[0] == "I": 
            self.newImgI(int(cmd[1]), int(cmd[2]))
        if cmd[0] == "C":
            self.clearC()
        if cmd[0] == "L":
            self.locationL(int(cmd[1]), int(cmd[2]), cmd[3])
        if cmd[0] == "V":
            self.vertV(int(cmd[1]), int(cmd[2]), int(cmd[3]), cmd[4])
        if cmd[0] == "H":
            self.horizH(int(cmd[1]), int(cmd[2]), int(cmd[3]), cmd[4])
        if cmd[0] == "K":
            self.rectK(int(cmd[1]), int(cmd[2]), int(cmd[3]), int(cmd[4]), cmd[5])
        if cmd[0] == "F":
            self.fillF(int(cmd[1]), int(cmd[2]), cmd[3])
        if cmd[0] == "S":
            self.saveS(cmd[1])
        
    def newImgI(self, M, N):
        #return [["0"] * int(cmd[1])] * int(cmd[2]) 
        #creates duplicates of same list(all elements in column affected)
        self.img = []
        for i in range(0,N):
            self.img.append(["O" for x in range(0,M)])

    def locationL(self, X, Y, C):
        self.img[Y - 1][X - 1] = C
        
    def clearC(self):
        self.img = [["O" for x in self.img[0]] for x in self.img]
        
    def vertV(self, X, Y1, Y2, C):
        
        for y in range((min(Y2,Y1) - 1 ), (max(Y1,Y2))):
            self.img[y][X-1] = C  

    def horizH(self, X1, X2, Y, C):
        lower = min(X2,X1) - 1
        upper = max(X1,X2) - 1
        
        """for x in range(lower, upper + 1):
        
            self.img[Y-1][x] = C #* (upper + 1) - lower)"""
        

        self.img[Y-1][lower:upper + 1] = C * ((upper - lower) + 1)



    def rectK(self, X1, Y1, X2, Y2, C):
    
        lower = min(X1,X2) - 1
        upper = max(X2,X1) - 1
        
        for y in range(min(Y1,Y2) - 1, max(Y2,Y1)):
        
            self.img[y][lower:upper + 1] = C * ((upper - lower) + 1)
    
    def fillF(self, X, Y, C):
        
        maxX = len(self.img[0])
        maxY = len(self.img)
        
        origX = X - 1
        origY = Y - 1
        
        oldColor = self.img[origY][origX]
        
        
        if oldColor != C:
            
            queueCur = 0
            
            queue = []
            visited = set()
            
            #img[origY][origX] = newColor
            queue.append((origX,origY))
            visited.add((origX,origY))
            
            nextX = 0
            nextY = 0
            
            self.img[queue[queueCur][1]][queue[queueCur][0]] = C
            if (queue[queueCur][0], queue[queueCur][1]) not in visited: 
                visited.add((queue[queueCur][0], queue[queueCur][1]))
         
        
            if (queue[queueCur][0], queue[queueCur][1]) not in visited: 
                visited.add((queue[queueCur][0], queue[queueCur][1]))
         
            while queueCur < len(queue):
                for i in range(-1,2):
                    for j in range(-1,2):
                            nextX = queue[queueCur][0] + j
                            nextY = queue[queueCur][1] + i
                            
                            if nextX in range(0, maxX) and nextY in range(0,maxY):
                                if self.img[nextY][nextX] == oldColor:
                                    self.img[nextY][nextX] = C
                                    if (nextX,nextY) not in visited:
                                        queue.append((nextX,nextY))
                queueCur += 1


    def saveS(self,Name):
        print(Name)
        
        for x in self.img:
            for y in x:
                print(y, end="")
            print("")

# Class End            
    
if __name__ == "__main__":
    img = Image()
    #with open('pc165_inputs.txt') as f:
    for line in stdin:#f: 
        #print(line)
        img.command([x for x in line.split()])
    del img
    print()

File: Chapter 1/5. Graphical Editor/pc165.cpp
#include <iostream>
#include <vector>
#include <string>
#include <set>
#include <algorithm>
#include <sstream>
#include <fstream>


class Image {
public:
    Image() {}

    void command(const std::vector<std::string>& cmd) {
        if (cmd[0] == "I") {
            newImgI(std::stoi(cmd[1]), std::stoi(cmd[2]));
        }
        if (cmd[0] == "C") {
            clearC();
        }
        if (cmd[0] == "L") {
            locationL(std::stoi(cmd[1]), std::stoi(cmd[2]), cmd[3]);
        }
        if (cmd[0] == "V") {
            vertV(std::stoi(cmd[1]), std::stoi(cmd[2]), std::stoi(cmd[3]), cmd[4]);
        }
        if (cmd[0] == "H") {
            horizH(std::stoi(cmd[1]), std::stoi(cmd[2]), std::stoi(cmd[3]), cmd[4]);
        }
        if (cmd[0] == "K") {
            rectK(std::stoi(cmd[1]), std::stoi(cmd[2]), std::stoi(cmd[3]), std::stoi(cmd[4]), cmd[5]);
        }
        if (cmd[0] == "F") {
            fillF(std::stoi(cmd[1]), std::stoi(cmd[2]), cmd[3]);
        }
        if (cmd[0] == "S") {
            saveS(cmd[1]);
        }
    }

private:
    std::vector<std::vector<std::string>> img;

    void newImgI(int M, int N) {
        img.clear();
        img.resize(N, std::vector<std::string>(M, "O"));
    }

    void locationL(int X, int Y, const std::string& C) {
        img[Y - 1][X - 1] = C;
    }

    void clearC() {
        for (auto& row : img) {
            std::fill(row.begin(), row.end(), "O");
        }
    }

    void vertV(int X, int Y1, int Y2, const std::string& C) {
        for (int y = std::min(Y2, Y1) - 1; y < std::max(Y1, Y2); ++y) {
            img[y][X - 1] = C;
        }
    }

    void horizH(int X1, int X2, int Y, const std::string& C) {
        int lower = std::min(X2, X1) - 1;
        int upper = std::max(X1, X2) - 1;
        std::fill(img[Y - 1].begin() + lower, img[Y - 1].begin() + upper + 1, C);
    }

    void rectK(int X1, int Y1, int X2, int Y2, const std::string& C) {
        int lower = std::min(X1, X2) - 1;
        int upper = std::max(X2, X1) - 1;
        for (int y = std::min(Y1, Y2) - 1; y < std::max(Y2, Y1); ++y) {
            std::fill(img[y].begin() + lower, img[y].begin() + upper + 1, C);
        }
    }

    void fillF(int X, int Y, const std::string& C) {
        int maxX = img[0].size();
        int maxY = img.size();
        int origX = X - 1;
        int origY = Y - 1;
        std::string oldColor = img[origY][origX];

        if (oldColor != C) {
            std::vector<std::pair<int, int>> queue;
            std::set<std::pair<int, int>> visited;

            queue.emplace_back(origX, origY);
            visited.insert({origX, origY});
            img[origY][origX] = C;

            while (!queue.empty()) {
                auto [curX, curY] = queue.back();
                queue.pop_back();

                for (int i = -1; i <= 1; ++i) {
                    for (int j = -1; j <= 1; ++j) {
                        if (std::abs(i) + std::abs(j) == 1) { // Only adjacent cells
                            int nextX = curX + j;
                            int nextY = curY + i;

                            if (nextX >= 0 && nextX < maxX && nextY >= 0 && nextY < maxY) {
                                if (img[nextY][nextX] == oldColor) {
                                    img[nextY][nextX] = C;
                                    if (visited.insert({nextX, nextY}).second) {
                                        queue.emplace_back(nextX, nextY);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    void saveS(const std::string& Name) {
        std::cout << Name << std::endl;
        for (const auto& row : img) {
            for (const auto& cell : row) {
                std::cout << cell;
            }
            std::cout << std::endl;
        }
    }
};

int main() {
    Image img;
    std::string line;

    
    while (std::getline(std::cin, line)) {
        if (line != "X") {
            std::vector<std::string> cmd;
            std::istringstream iss(line);
            std::string token;
            while (iss >> token) {
                cmd.push_back(token);
            }
            img.command(cmd);
        } else {
            //std::cout << std::endl;
            break;
        }
    }
    
    /*
    std::ifstream file("pc165_inputs.txt");
    while (std::getline(file, line)) {
        if (line != "X") {
            std::vector<std::string> cmd;
            std::istringstream iss(line);
            std::string token;
            while (iss >> token) {
                cmd.push_back(token);
            }
            img.command(cmd);
        } else {
            //std::cout << std::endl;
            break;
        }
    }*/

    return 0;
}